### ISSUE TITLE
The options stack does not get reset after an `Error` in a test file

### ISSUE THREAD

**Issue by wilfwilson** [gap-system/gap#4510]
Let us create a function whose behaviour depends on a value option:
```gap
WilfNumber := function(G)
  if not (IsGroup(G) and IsSimple(G)) then
    Error("needs to be a simple group!");
  fi;
  if ValueOption("blah") <> fail then
    Error("This line should not be executed by any test");
  fi;
  return 5;
end;;
```
If I create a test file `option.tst` with the following contents:
```gap
gap> START_TEST("option.tst");
gap> WilfNumber(AlternatingGroup(4) : blah);
Error, needs to be simple!
gap> WilfNumber(AlternatingGroup(5));
5
gap> STOP_TEST("option.tst");
```
then this is the result of running that test:
```gap
gap> Test("option.tst");
########> Diff in option.tst:4
# Input is:
WilfNumber(AlternatingGroup(5));
# Expected output:
5
# But found:
Error, This line should not be executed by any test
########
option.tst
msecs: 1
false
```
This shows that `ValueOption("blah")` remains set as `true` for the second call of `WilfNumber`. Indeed, this is what I get at the GAP prompt after running that test file:
```
gap> ValueOption("blah");
true
```
I came across this while helping to review https://github.com/digraphs/Digraphs/pull/382 (there we use `ErrorNoReturn` instead of `Error`, but otherwise it's essentially the same problem. The 'wrongly set' option meant that an inappropriate algorithm was being used in a method, which required too much memory in the particular case, and this caused GAP to crash.)

Perhaps the options stack was never intended to get reset after an error in a test file, but I think it should be. In my mind, when a test file continues beyond an error, the test should do the same thing as if the user had typed "quit;" at the GAP prompt. In the latter case, things behave as I would have expected:
```
gap> WilfNumber(AlternatingGroup(4));
Error, needs to be simple! at *stdin*:14 called from
<function "WilfNumber">( <arguments> )
 called from read-eval loop at *stdin*:39
you can 'quit;' to quit to outer loop, or
you can 'return;' to continue
brk> quit;
#I  Options stack has been reset
gap> ValueOption("blah");
fail
```

**Comment by fingolfin** [comment]
This is clearly a bug, agreed

**Comment by ChrisJefferson** [comment]
Was just tracing what current happens:

We have a function `OnQuit`, which does various cleanups when we exit a break loop. That function clears the OptionsStack (which actually isn't perfect, as someone could have tried to set an option globally, and it would be cleared, but that's current behaviour).

That function isn't getting called by testing. I'm not (yet) positive if the best option is to call `OnQuit` (which does some other cleanups), or just clean the OptionsStack. Probably do everything.

**Comment by fingolfin** [comment]
It's more complicated than that, I think, as IIRC OnQuit also prints the backtrace, which we want to suppress. Moreover, some packages add code to OnQuit, too

**Comment by ChrisJefferson** [comment]
Yes. I'm looking at this -- I'm trying out just copying the options reset into `test`. This also requires adding an `OnBreak` to test, so we know if the break loop was emptied.

**Comment by fingolfin** [comment]
I think it is weird that OnQuit flushes the whole option stack anyway, precisely because it messed with any global options the user may have pushed. It would make more sense if it restored them to the state before the top-level call which ultimately lead to the error. This is of course harder to implement (may need help from the kernel).

BTW I "love" how the `OnQuit` documentation says 
> Currently, OnQuit is not advertised, since exception handling may make it obsolete.

**Comment by fingolfin** [comment]
This was resolved by @ThomasBreuer in PR #6355 and thus in GAP 4.16.0
